### ISSUE TITLE
Handle Heat client exceptions

### ISSUE THREAD

**Issue by arbrandes** [hastexo/hastexo-xblock#82]
Handle certain Heat client exceptions so that the multi-provider algorithm continues execution.

504 on create:

```
[service_variant=lms][celery.wor ERROR [app1 18960] [log.py:282] - Task hastexo.tasks.LaunchStackTask[397bb0b0-fd59-4101-8100-776031e02fd4] raised unexpected: HTTPException() Traceback (most recent call last): File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 240, in trace_task R = retval = fun(*args, **kwargs) File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 438, in __protected_call__ return self.run(*args, **kwargs) File "/edx/app/edxapp/venvs/edxapp/src/hastexo-xblock/hastexo/tasks.py", line 69, in run stack_data = self.launch_stack() File "/edx/app/edxapp/venvs/edxapp/src/hastexo-xblock/hastexo/tasks.py", line 165, in launch_stack stack_data = self.try_all_providers() File "/edx/app/edxapp/venvs/edxapp/src/hastexo-xblock/hastexo/tasks.py", line 213, in try_all_providers stack_data = self.try_provider(provider["name"]) File "/edx/app/edxapp/venvs/edxapp/src/hastexo-xblock/hastexo/tasks.py", line 322, in try_provider heat_stack = self.create_stack(heat_c, provider, env) File "/edx/app/edxapp/venvs/edxapp/src/hastexo-xblock/hastexo/tasks.py", line 382, in create_stack parameters=parameters File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/heatclient/v1/stacks.py", line 172, in create data=kwargs, headers=headers) File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/keystoneauth1/adapter.py", line 357, in post return self.request(url, 'POST', **kwargs) File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/heatclient/common/http.py", line 318, in request raise exc.from_response(resp) HTTPException: ERROR: <html><body><h1>504 Gateway Time-out</h1> The server didn't respond in time. </body></html>
```

404 on cleanup:

```
[service_variant=lms][celery.wor ERROR [app2 19839] [log.py:282] - Task hastexo.tasks.LaunchStackTask[67b1ee7c-88db-45ba-bfd1-e9643dc79ded] raised unexpected: HTTPNotFound() Traceback (most recent call last): File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 240, in trace_task R = retval = fun(*args, **kwargs) File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 438, in __protected_call__ return self.run(*args, **kwargs) File "/edx/app/edxapp/venvs/edxapp/src/hastexo-xblock/hastexo/tasks.py", line 90, in run self.cleanup_stack(e) File "/edx/app/edxapp/venvs/edxapp/src/hastexo-xblock/hastexo/tasks.py", line 475, in cleanup_stack heat_c.stacks.delete(stack_id=self.stack_name) File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/heatclient/v1/stacks.py", line 202, in delete self._delete("/stacks/%s" % stack_id) File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/heatclient/openstack/common/apiclient/base.py", line 228, in _delete return self.client.delete(url) File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/keystoneauth1/adapter.py", line 366, in delete return self.request(url, 'DELETE', **kwargs) File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/heatclient/common/http.py", line 318, in request raise exc.from_response(resp) HTTPNotFound: ERROR: The Stack (cc216_201811-summit_7c0a7f441fb834084bc30967acd6b0c2) could not be found.
```

409 on create:

```
[service_variant=lms][celery.wor ERROR [app2 19839] [log.py:282] - Task hastexo.tasks.LaunchStackTask[e80dce81-a8e4-4d5d-87cd-e1ed793f171a] raised unexpected: HTTPConflict() Traceback (most recent call last): File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 240, in trace_task R = retval = fun(*args, **kwargs) File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 438, in __protected_call__ return self.run(*args, **kwargs) File "/edx/app/edxapp/venvs/edxapp/src/hastexo-xblock/hastexo/tasks.py", line 69, in run stack_data = self.launch_stack() File "/edx/app/edxapp/venvs/edxapp/src/hastexo-xblock/hastexo/tasks.py", line 165, in launch_stack stack_data = self.try_all_providers() File "/edx/app/edxapp/venvs/edxapp/src/hastexo-xblock/hastexo/tasks.py", line 213, in try_all_providers stack_data = self.try_provider(provider["name"]) File "/edx/app/edxapp/venvs/edxapp/src/hastexo-xblock/hastexo/tasks.py", line 322, in try_provider heat_stack = self.create_stack(heat_c, provider, env) File "/edx/app/edxapp/venvs/edxapp/src/hastexo-xblock/hastexo/tasks.py", line 382, in create_stack parameters=parameters File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/heatclient/v1/stacks.py", line 172, in create data=kwargs, headers=headers) File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/keystoneauth1/adapter.py", line 357, in post return self.request(url, 'POST', **kwargs) File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/heatclient/common/http.py", line 318, in request raise exc.from_response(resp) HTTPConflict: ERROR: The Stack (cc216_201811-summit_d83dbb95abfb3163539ee5519d9d3154) already exists.
```

500 when getting:
```
[service_variant=lms][celery.wor ERROR [app0 19264] [log.py:282] - Task hastexo.tasks.LaunchStackTask[dc4e5f08-2d12-4570-a6ce-c560d0f878a2] raised unexpected: HTTPInternalServerError() Traceback (most recent call last): File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 240, in trace_task R = retval = fun(*args, **kwargs) File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 438, in __protected_call__ return self.run(*args, **kwargs) File "/edx/app/edxapp/venvs/edxapp/src/hastexo-xblock/hastexo/tasks.py", line 69, in run stack_data = self.launch_stack() File "/edx/app/edxapp/venvs/edxapp/src/hastexo-xblock/hastexo/tasks.py", line 165, in launch_stack stack_data = self.try_all_providers() File "/edx/app/edxapp/venvs/edxapp/src/hastexo-xblock/hastexo/tasks.py", line 213, in try_all_providers stack_data = self.try_provider(provider["name"]) File "/edx/app/edxapp/venvs/edxapp/src/hastexo-xblock/hastexo/tasks.py", line 322, in try_provider heat_stack = self.create_stack(heat_c, provider, env) File "/edx/app/edxapp/venvs/edxapp/src/hastexo-xblock/hastexo/tasks.py", line 397, in create_stack heat_stack = heat_c.stacks.get(stack_id=heat_stack.id) File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/heatclient/v1/stacks.py", line 274, in get resp = self.client.get('/stacks/%s' % stack_id, **kwargs) File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/keystoneauth1/adapter.py", line 351, in get return self.request(url, 'GET', **kwargs) File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/heatclient/common/http.py", line 318, in request raise exc.from_response(resp) HTTPInternalServerError: ERROR: Timed out waiting for a reply to message ID f46f51706e4f48c9889f5a62b6c6c5b0
```
```

**Comment by arbrandes** [comment]
Proposed fix in https://github.com/hastexo/hastexo-xblock/pull/83

**Comment by fghaas** [comment]
Fixed in #83, closing.
